### PR TITLE
add cache-folder config option for setting cache folder location

### PR DESCRIPTION
**Pull request by superxp1412**
**Summary**

Yarn need a elegant way to config cache folder's directory.   
This pull request is an answer to: [#1316](https://github.com/yarnpkg/yarn/issues/1316) & [#1211](https://github.com/yarnpkg/yarn/issues/1211)

With this feature we can define the location of yarn's cache folder:
e.g.
`yarn config set cache-folder ${path}` 

And the settings would be defined inside of .yarnrc file like:
`cache-folder "${path}"`

Besides, new configurations for cache-folder can also been verified by command:
`yarn cache dir`

@samccone

### CODE REVIEW

**Comment by superxp1412** [comment]
Yes, I'll try to add test for this feature.

**Comment by bestander** [comment]
@superxp1412, you should just use `install` tests to verify where the cache is coming from https://github.com/yarnpkg/yarn/blob/master/__tests__/commands/install.js#L726.
You can pass options and flags to the install command to configure where the cache is stored.

File: src/config.js [hunk 1]
```diff
@@ -203,6 +200,12 @@ export default class Config {
       cert: String(opts.cert || this.getOption('cert') || ''),
       key: String(opts.key || this.getOption('key') || ''),
     });
+
+    //init & create cacheFolder, tempFolder
+    this.cacheFolder = String(opts.cacheFolder || this.getOption('cache-folder') || constants.MODULE_CACHE_DIRECTORY);
```
Comment (bestander) [inline src/config.js]: The only test needed is that all 3 options may setup cache folder location and they have different priority

Comment (superxp1412) [inline src/config.js]: I've already submitted the unit test for the priorities for the options. Please review it.
@bestander

**Comment by bestander** [comment]
@superxp1412 any idea why tests failed?
Maybe Circle/Windows does not have access to the config folder.

**Comment by superxp1412** [comment]
I think the test fails may caused by concurrently read/write .yarnrc file. There's a chance to reproduce that issue. 
Now, I've fixed it by changing the test running methods been run in sequence.
@bestander

**Comment by mvestergaard** [comment]
This seems fairly broken atm.
This line in `src/config.js@205`: `this.cacheFolder = String(opts.cacheFolder || this.getOption('cache-folder') || constants.MODULE_CACHE_DIRECTORY);`

Sets `cacheFolder` to `set_config_folder_dir` because of `this.getOption('cache-folder')`. I'm not entirely sure where that value comes from, but it ends up creating that folder in your working directory, and using that folder as cache.

**Comment by mvestergaard** [comment]
It also creates a `flag_config_folder_dir` inside your yarn working directory whenever you run tests.

**Comment by superxp1412** [comment]
Hi @mvestergaard , @bestander 
Yes, the tests is not sandbox well. 
I promise to fix it by another pull request soon.

**Comment by bestander** [comment]
@mvestergaard, could you raise an issue to track down your problem and ping me?
Can you send a PR with a fix for your situation?

**Comment by mvestergaard** [comment]
@bestander sure i can add an issue if you'd like. I already created the PR #2144 with a fix

**Comment by bestander** [comment]
Thanks, no need for issue if the PR is there :)

**Comment by buschtoens** [comment]
I know now it's to late, but `folder` is a rather unconventional name. It really should have been `dir` or `directory`.

https://superuser.com/questions/169457/directory-vs-folder

**Comment by bestander** [comment]
@buschtoens, we use word "folder" in 85 places, so this might be a big change.
If you feel strong about it and willing to do a codemod, please send an RFC, we can discuss the pros and cons.

**Comment by buschtoens** [comment]
Thanks for taking note, @bestander! :smile: 

I primarily commented to vent for a bit, as I passed `--cache-dir` instead of `--cache-folder`. Sorry. :sweat_smile: 

Since this is already implemented, taught and in people's heads, changing it probably won't be worth the trouble.

**Comment by bestander** [comment]
I totally understand your frustration.
How about all the possible ways to clear cache in a tool: --no-cache,
--clear-cache, clean, cache clean, --reset-cache?

**Pull request merged**